JavaScript development/Frameworks/React.md

```diff
@@ -0,0 +1,598 @@
+
+In class components:
+- always use prop validation
+- use **static member** for declaring prop types and default props (if your transpiler doesn't support static members, set them after class declaration and then export the class; See: ["Exporting Components" section](#exporting-components))
+- use *defaultProps* if you need to set default values
+- prop types shoud be as detailed as possible 
+
+``` jsx
+export default class ComponentName extends React.Component {
+  static propTypes = {
+    heading: PropTypes.string,
+    users: PropTypes.arrayOf(
+      PropTypes.shape({
+        id: PropTypes.string.isRequired,
+        firstName: PropTypes.string.isRequired,
+        lastName: PropTypes.string.isRequired,
+        age: PropTypes.number
+      })
+    ).isRequired,
+  };
+
+  static defaultProps = {
+    heading: 'Default Heading',
+  };
+
+  // ...
+
+  render() {
+    const {name} = this.props;
+
+    return (
+      <div>{name}</div>
+    );
+  }
+}
+```
+
+In stateless components, use PropTypes for type checking but use native JS for setting default values.
+
+``` jsx
+function ComponentName({id, name = 'John Doe'}) {
+  // ...
+};
+
+ComponentName.propTypes = {
+  id: PropTypes.number.isRequired,
+  name: PropTypes.string
+}
+```
+
+### Using props
+
+If you are using more than one property in a method, preffer accessing them with **object destructuring**.
+
+``` jsx
+export class ComponentName extends React.Component {
+  // ...
+  render() {
+    const {id, name, onClickHandler} = this.props;
+    // ...
+  }
+}
+```
+
+### Importing React
+``` jsx
+import React from 'react';
+
+export class ComponentName extends React.Component {
+  // ...
+}
+```
+
+### Exporting Components
+- each React component file should have a single exported component
+- don't use default exports, always use named exports instead
+ 
+Exporting class components:
+``` jsx
+// bad
+export default class ComponentName extends React.Component {
+  // ...
+}
+
+// good
+export class ComponentName extends React.Component {
+  // ...
+}
+```
+
+Exporting class components when you need to declare static members afterwards:
+``` jsx
+// exporting when static members are not supported by the transpiler:
+class ComponentName extends React.Component {
+  // ...
+}
+
+ComponentName.propTypes = {
+  // ..
+};
+
+export {ComponentName};
+```
+
+Exporting functional components:
+``` jsx
+// stateless components with no prop validation
+export function FunctionalComponent() {
+  // ...
+};
+
+
+// stateless components with static prop validation
+function FunctionalComponent(/* props */) {
+  // ...
+}
+
+FunctionalComponent.propTypes = {
+  // ...
+};
+
+export {FunctionalComponent};
+```
+
+### Rendering Arrays of Data
+
+For iterating arrays:
+- use **map** method to render items
+- you should have a unique key for each element
+- **don't use array's index as a key**, unless you have no choice (no unique property in each element of the array)
```

Comment:
I would also add that the elements need to keep their positions (constant indexes). If not, this causes some very unpredictable bugs.